agents/agents-ext/src/commonMain/kotlin/ai/koog/agents/ext/agent/SubgraphWithRetry.kt

```diff
@@ -0,0 +1,163 @@
+package ai.koog.agents.ext.agent
+
+import ai.koog.agents.core.agent.context.AIAgentContextBase
+import ai.koog.agents.core.agent.entity.ToolSelectionStrategy
+import ai.koog.agents.core.agent.entity.createStorageKey
+import ai.koog.agents.core.dsl.builder.AIAgentSubgraphBuilderBase
+import ai.koog.agents.core.dsl.builder.AIAgentSubgraphDelegate
+import ai.koog.agents.core.dsl.builder.forwardTo
+
+/**
+ * Represents the result of [subgraphWithRetry].
+ *
+ * @param output The result of the subgraph operation.
+ * @param success A boolean indicating whether the action was successful.
+ * @param retryCount The number of retries attempted.
+ */
+public data class RetrySubgraphResult<Output>(
+    val output: Output,
+    val success: Boolean,
+    val retryCount: Int,
+)
+
+/**
+ * Creates a subgraph with retry mechanism, allowing a specified action subgraph to be retried multiple
+ * times until a given condition is met or the maximum number of retries is reached.
+ *
+ * @param condition A function that evaluates whether the output meets the desired condition.
+ * @param maxRetries The maximum number of allowed retries. Must be greater than 0.
+ * @param toolSelectionStrategy The strategy used to select a tool for executing the action.
+ * @param name The optional name of the subgraph.
+ * @param defineAction A lambda defining the action subgraph to perform within the retry subgraph.
+ */
+public fun <Input : Any, Output> AIAgentSubgraphBuilderBase<*, *>.subgraphWithRetry(
+    condition: suspend (Output) -> Boolean,
+    maxRetries: Int,
+    toolSelectionStrategy: ToolSelectionStrategy = ToolSelectionStrategy.ALL,
+    name: String? = null,
+    defineAction: AIAgentSubgraphBuilderBase<Input, Output>.() -> Unit,
+): AIAgentSubgraphDelegate<Input, RetrySubgraphResult<Output>> {
+    require(maxRetries > 0) { "maxRetries must be greater than 0" }
+
+    return subgraph(name = name) {
+        val retriesKey = createStorageKey<Int>("${name}_retires")
```

Comment:
Not sure if we should use storage here, as it can probably rewrite user's key. Maybe we can reserve some name for us and come up with naming convention for them (e.g., \_\_name\_\_)

Reply:
As a workaround here, we can operate with the input wrapped into the class which also contains the retries info (something like `RetrySubgraphResult`)

Reply:
And also add initial input and context there

Reply:
> Not sure if we should use storage here, as it can probably rewrite user's key

We won't rewirte this, because the key in the storage is not a name (it's here only for convenience), but the object. So, every time you call `createStorageKey`, a unique object (key) is created, even if the names are the same. So it's safe to do it like this, we won't overwrite any existing data.